File: dutch_national_flag.py
# Dutch national flag problem

def sort_012(input_list):
    """
    Given an input array consisting on only 0, 1, and 2, sort the array in a single traversal.

    Args:
       input_list(list): List to be sorted
    """
    left_index = 0
    right_index = len(input_list) - 1
    center_index = 0

    while center_index <= right_index:
        # if we find a 0, push it to the left side
        if input_list[center_index] == 0:
            input_list[center_index] = input_list[left_index]
            input_list[left_index] = 0 
            left_index += 1
            center_index += 1
        # if we find a 2, push it to the right side
        elif input_list[center_index] == 2:
            input_list[center_index] = input_list[right_index]
            input_list[right_index] = 2
            right_index -= 1
        # not 0, and not 2? must be a 1, keep going
        else:
            center_index += 1

    return input_list



def test_function(test_case):
    sorted_array = sort_012(test_case)
    
    if sorted_array == sorted(test_case):
        print("Pass")
    else:
        print("Fail")

# test cases
test_function([0, 0, 2, 2, 2, 1, 1, 1, 2, 0, 2])
test_function([2, 1, 2, 0, 0, 2, 1, 0, 1, 0, 0, 2, 2, 2, 1, 2, 0, 0, 0, 2, 1, 0, 2, 0, 0, 1])
test_function([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2])
test_function([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
test_function([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
test_function([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])
test_function([0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2])
test_function([])
test_function([2, 1, 0])

File: arrays_and_strings/dutch_national_flag.php
<?php

/*
Dutch national flag
*/

class Solution
{
    public function dnf($nums) {
        $left = 0;
        $right = count($nums) - 1;
        $i = 0;

        while ($i <= $right) {
            if ($nums[$i] == 0) {
                $nums[$i] = $nums[$left];
                $nums[$left] = 0;
                $left += 1;
                $i += 1;
            }
            elseif ($nums[$i] == 2) {
                $nums[$i] = $nums[$right];
                $nums[$right] = 2;
                $right -= 1;
            } else {
                $i += 1;
            }
        }
        
        return $nums;
    }
}


$s = new Solution();

$nums = [2, 1, 2, 0, 0, 2, 1, 0, 1, 0, 0, 2, 2, 2, 1, 2, 0, 0, 0, 2, 1, 0, 2, 0, 0, 1];
$sorted = $s->dnf($nums);

echo implode(", ", $sorted);

?>